### PR TITLE
lib: bin: lwm2m_carrier: update to version 0.22.0

### PR DESCRIPTION
**Pull request by nordic-hani**
Update library to version 0.22.0

### CODE REVIEW

File: doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst [hunk 1]
```diff
@@ -9,6 +9,43 @@ Changelog
...
+| Library size            |               |            |
+| (reference application) |               |            |
++-------------------------+---------------+------------+
+
+Changes
+=======
+
+* Added a new event :c:macro:`LWM2M_CARRIER_DEFERRED_SIM_MSISDN`.
+
+  * This event can trigger on devices with new a SIM card which has not been registered on the network yet.
```
Comment (peknis) [inline doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst:42]:
```suggestion
  * This event can trigger on devices with a new SIM card that has not been registered on the network yet.
```

File: doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst [hunk 2]
```diff
@@ -9,6 +9,43 @@ Changelog
...
+    (Therefore it has not received the phone number needed for device management).
```
Comment (peknis) [inline doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst:43]:
```suggestion
    (Therefore, it has not received the phone number needed for device management).
```

File: doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst [hunk 3]
```diff
@@ -9,6 +9,43 @@ Changelog
...
+* Removed dependency on the deprecated AT Command and AT notification libraries from the glue layer.
```
Comment (peknis) [inline doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst:44]:
```suggestion
* Removed dependency on the deprecated AT command and AT notification libraries from the glue layer.
```

File: doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst [hunk 4]
```diff
@@ -9,6 +9,43 @@ Changelog
...
+* Added dependency on the AT monitor library in the glue layer.
+* Changed the :c:func:`lwm2m_os_lte_link_up` function to perform asynchronous connect.
```
Comment (peknis) [inline doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst:46]:
```suggestion
* Changed the :c:func:`lwm2m_os_lte_link_up` function to perform an asynchronous connect.
```

File: doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst [hunk 5]
```diff
@@ -9,6 +9,43 @@ Changelog
...
+* Removed the following unused functions from glue layer: ``lwm2m_os_sec_ca_chain_exists()``, ``lwm2m_os_sec_ca_chain_cmp()``, ``lwm2m_os_sec_ca_chain_write()``.
```
Comment (peknis) [inline doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst:47]:
```suggestion
* Removed the following unused functions from the glue layer: ``lwm2m_os_sec_ca_chain_exists()``, ``lwm2m_os_sec_ca_chain_cmp()``, ``lwm2m_os_sec_ca_chain_write()``.
```

File: doc/nrf/libraries/bin/lwm2m_carrier/app_integration.rst [hunk 6]
```diff
@@ -179,7 +179,23 @@ Following are the various LwM2M carrier library events:
...
+    * :c:macro:`LWM2M_CARRIER_ERROR_BOOTSTRAP` - This error is generated during the bootstrap procedure.
+
+      +--------------------------------------------------------+-----------------------------------------------------------------------------+--------------------------------------------------+
+      | Errors                                                 | More information                                                            | Recovery                                         |
+      |                                                        |                                                                             |                                                  |
+      +========================================================+=============================================================================+==================================================+
+      | Retry limit for connecting to the bootstrap            | Common reason for this failure can be incorrect URI or PSK,                 | Library will retry after next device reboot      |
```
Comment (peknis) [inline doc/nrf/libraries/bin/lwm2m_carrier/app_integration.rst:188]:
```suggestion
      | Retry limit for connecting to the bootstrap            | Common reason for this failure can be incorrect URI or PSK,                 | Library will retry after next device reboot.      |
```

File: doc/nrf/libraries/bin/lwm2m_carrier/app_integration.rst [hunk 7]
```diff
@@ -179,7 +179,23 @@ Following are the various LwM2M carrier library events:
...
+      | server has been reached (``-ETIMEDOUT``).              | or the server is unavailable (for example temporary network issues).        |                                                  |
+      |                                                        | If this error persists, contact your carrier.                               |                                                  |
+      +--------------------------------------------------------+-----------------------------------------------------------------------------+--------------------------------------------------+
+      | Failure to provision the PSK                           | If the link is forced up by the application during the bootstrap procedure  | Library will retry after 24 hours                |
```
Comment (peknis) [inline doc/nrf/libraries/bin/lwm2m_carrier/app_integration.rst:192]:
```suggestion
      | Failure to provision the PSK                           | If the link is forced up by the application during the bootstrap procedure  | Library will retry after 24 hours.                |
```

File: doc/nrf/libraries/bin/lwm2m_carrier/requirements.rst [hunk 8]
```diff
@@ -31,11 +31,9 @@ Below are some of the requirements and limitations of the application while runn
    * The LwM2M carrier library stores keys into the modem, which requires disconnecting from the LTE link and connecting to it.
    * The application must wait for the :c:macro:`LWM2M_CARRIER_EVENT_LTE_READY` event before using the LTE link.
 
-* The LwM2M carrier library uses the modem DFU socket and a TLS socket for FOTA.
+* The LwM2M carrier library uses the a TLS socket for FOTA.
```
Comment (peknis) [inline doc/nrf/libraries/bin/lwm2m_carrier/requirements.rst:34]:
```suggestion
* The LwM2M carrier library uses the TLS socket for FOTA.
```

File: doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst [hunk 9]
```diff
@@ -9,6 +9,43 @@ Changelog
...
+
++-------------------------+---------------+------------+
+|                         | Flash (Bytes) | RAM (Bytes)|
++-------------------------+---------------+------------+
+| Library size            |               |            |
+| (binary)                |               |            |
++-------------------------+---------------+------------+
+| Library size            |               |            |
+| (reference application) |               |            |
++-------------------------+---------------+------------+
```
Comment (peknis) [inline doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst:35]: I guess the sizes will be added to the table at some point.

**Review by peknis — CHANGES_REQUESTED** [review]

File: doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst (hunk 9 above)
Comment (nordic-hani) [reply doc/nrf/libraries/bin/lwm2m_carrier/CHANGELOG.rst:35]: Yes, closer to merge :)

File: doc/nrf/libraries/bin/lwm2m_carrier/app_integration.rst [hunk 10]
```diff
@@ -20,7 +20,7 @@ It provides an abstraction of the following modules:
   * :ref:`lte_lc_readme`
   * :ref:`lib_download_client`
   * :ref:`modem_key_mgmt`
-  * :ref:`at_cmd_parser_readme`
+  * :ref:`at_monitor_readme`
```
Comment (57300) [inline doc/nrf/libraries/bin/lwm2m_carrier/app_integration.rst:23]: AT monitor appears on the list twice.

File: samples/nrf9160/lwm2m_carrier/prj.conf [hunk 11]
```diff
@@ -32,9 +33,9 @@ CONFIG_SMS=y
 CONFIG_DOWNLOAD_CLIENT=y
 CONFIG_DOWNLOAD_CLIENT_MAX_FILENAME_SIZE=230
 
-# AT commands
-CONFIG_AT_CMD_PARSER=y
+# AT Command driver interface with Modem
```
Comment (57300) [inline samples/nrf9160/lwm2m_carrier/prj.conf:36]: `# AT monitor`?

File: lib/nrf_modem_lib/nrf_modem_os.c [hunk 12]
```diff
@@ -230,8 +230,7 @@ bool nrf_modem_os_is_in_isr(void)
 	return k_is_in_isr();
 }
 
-#define NRF_MODEM_OS_SEM_MAX 3
-static struct k_sem nrf_modem_os_sems[NRF_MODEM_OS_SEM_MAX];
+static struct k_sem nrf_modem_os_sems[NRF_MODEM_OS_NUM_SEM_REQUIRED];
```
Comment (stig-bjorlykke) [inline lib/nrf_modem_lib/nrf_modem_os.c:233]: Are the changes in `nrf_modem_os.c` related to the lwm2m_carrier update?

File: doc/nrf/libraries/bin/lwm2m_carrier/app_integration.rst [hunk 13]
```diff
@@ -179,7 +179,23 @@ Following are the various LwM2M carrier library events:
...
+      +--------------------------------------------------------+-----------------------------------------------------------------------------+--------------------------------------------------+
+      | Failure to provision the PSK                           | If the link is forced up by the application during the bootstrap procedure  | Library will retry after 24 hours.               |
+      | needed for the bootstrap procedure.                    | the error will be ``-EACCES``. Verify that the LwM2M library is controlling |                                                  |
+      |                                                        | the link until the :c:macro:`LWM2M_CARRIER_EVENT_LTE_READY` event is sent.  |                                                  |
+      +--------------------------------------------------------+-----------------------------------------------------------------------------+--------------------------------------------------+
+      | Failure to fetch MSISDN or ICCID values (``-EFAULT``). | These values are received from the network.                                 | Library will retry upon next network connection. |
```
Comment (stig-bjorlykke) [inline doc/nrf/libraries/bin/lwm2m_carrier/app_integration.rst:196]: `MSISDN` will be received from the network for some carriers.
`ICCID` is fetched from SIM.

File: lib/nrf_modem_lib/nrf_modem_os.c (hunk 12 above)
Comment (nordic-hani) [reply lib/nrf_modem_lib/nrf_modem_os.c]: We were just including the commit(s) from libmodem PR in our PR until the libmodem PR got merged. Rebased not that løibmodem v1.5.0 has been merged.

**Comment by nordic-hani** [comment]
Removed changelog entry, agreed with @umapraseeda to add the text directly to release notes instead.

**Review by peknis — APPROVED** [review]

**Review by lucasvdp — APPROVED** [review]

**Review by 57300 — APPROVED** [review]

**Review by eibe85 — APPROVED** [review]

**Review by stig-bjorlykke — APPROVED** [review]

**Review by Eoin-McMahon — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**